Alternate more flexible code owners mechanism, soon to avoid mass pings

infinisil opened the pull request:
## Description of changes

These damn mass pings are annoying, let's fix it!

This PR introduces an alternate mechanism of doing effectively the same as GitHub's native [CODEOWNERS feature](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners), but with some significant advantages:
- No reviews will be requested for PRs that target the wrong base branch.
- There is no need for user/team to have write access to be requested for reviews.
- Invalidity of the code owners file fails CI

This PR still runs the native CODEOWNERS together with the alternative mechanism, so that we can run it for some time to confirm that it works correctly. Once confirmed, we'll be able to easily turn off the native CODEOWNERS and just rely on the alternate mechanism.

Note that this PR depended on https://github.com/NixOS/org/issues/31

## Tested
- [x] Codeowner pings can't be avoided by changing the code owners in the same PR: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/13
- [x] Can't set non-org members as code owners: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/14
- [x] Reviews can be requested from people without write access: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/15
- [x] Files without owners cause no problems: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/16
- [x] Staging merges cause no problems: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/18
- [x] Pings get prevented and a comment is posted when the base branch is wrong: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/21
- [x] Can't add code owners for non-existent files: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/20
- [x] Drafts don't trigger pings: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/22 (and the CI logs show who would be pinged)
- [x] Undrafting triggers pings: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/23

---

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

philiptaron commented:
cc @tie for bash wizardry and excellent reviews as well.

tie commented:
Bash code is high quality, although I’d avoid `cmd1 < <(cmd2)` because it does not respect errexit (`-e` flag). E.g.
```console
$ bash -e -u -o pipefail -O inherit_errexit -c 'true < <(false); echo oops'
oops
```

I don’t think I’ll have time to pick this up, but I’d probably rewrite scripts in Python with GitHub client libraries to avoid Bash. It’s much nicer to work with if available, even if `subprocess.run` calls are a bit more bulky for scripting 😅
```python
#!nix-shell -i python3 -p python3Packages.pygithub
```

drupol commented:
I would definitely use something else than Bash too.

infinisil commented:
If I have time to work on this again, I'd prefer to keep using bash, since the script part is pretty much done already. But if somebody else picks this up before that, I fully support a Python rewrite!

SuperSandro2000 commented:
This would only cover the use case to request reviews for changed files. We would loose the UI indicator that files are owned by a codeowners, the feature to block merges unless a codeowber approved (we're are not using that to much) and the validation in the UI.

We could split the codeowners file to those features for some entries back though.

Mic92 commented:
> This would only cover the use case to request reviews for changed files. We would loose the UI indicator that files are owned by a codeowners, the feature to block merges unless a codeowber approved (we're are not using that to much) and the validation in the UI.
> 
> We could split the codeowners file to those features for some entries back though.

For me that sounds like a good tradeoff if it reduces the noise I am currently getting as a code owner.

infinisil commented:
Good point @SuperSandro2000. I agree with @Mic92 that it's probably a good tradeoff still, but in the future we can also fully or partially implement those features:
- UI indicators: Because we're not limited by GitHub anymore, we can make owners declarations very local to files/dirs, e.g. an `owner: @foo` at the top of files, or a `CODEOWNERS` file in local directories
- Blocking merges: Can be done by a GitHub action
- UI validation: Can be done with a GitHub action too, I actually set this up for another repo before: https://github.com/NixOS/org/blob/da5ea7ed87d326568db245657ad8e00b8525a816/.github/workflows/ci.yml#L24-L56

Ideally all of this would be implemented by some GitHub action out there already, but I couldn't find anything that would work for our specific case.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-reaching-the-right-reviewers/51312/4

infinisil committed:
ci: Add default.nix with exposed pkgs

Allows reusing it in more places

infinisil commented:
Worked on this a bunch, and a bit more is needed, but it's looking pretty good! I tested:
- That it correctly requests reviews: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/3
- (Big!) Support for setting people without write access as a code owner! https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/5
- Test that it doesn't request reviews when the base branch is wrong: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/6

Eveeifyeve commented:
So is there still going to be still `lib.maintainers` or is nixpkgs switching to owners?

infinisil commented:
@Eveeifyeve Good question. For now we should definitely maintain the status quo. `meta.maintainers` does have advantages (such as being able to process it more easily, e.g. for detecting unmaintained packages at eval-time), but also disadvantages (it doesn't map cleanly to files). I can imagine a future in which we combine the best of both worlds.

Eveeifyeve commented:
> @Eveeifyeve Good question. `meta.maintainers` does have advantages (such as being able to process it more easily, e.g. for detecting unmaintained packages at eval-time), but also disadvantages (it doesn't map cleanly to files). I can imagine a future in which we combine the best of both worlds, but for now we should definitely maintain the status quo.

Yeah but I guess it makes sense.

infinisil committed:
ci: Add review request scripts

Also post a comment in case base branch is wrong
  This guides newcomers in how to smoothly handle the potentially scary
  situation of having thousands of commits listed in a PR.

  While CI shows the same, people might not even look at CI if the PR
  looks botched.

infinisil committed:
ci: Add codeowners validator

infinisil committed:
workflows/codeowners: init

infinisil committed:
OWNERS: placeholder init

infinisil commented:
This is ready imo! Can't be merged just yet due to it depending on https://github.com/NixOS/org/issues/31, but the code should be fine :)

cafkafk commented on ci/codeowners-validator/default.nix:
```diff
@@ -0,0 +1,31 @@
+  fetchpatch,
+}:
+buildGoModule {
+  name = "codeowners-validator";
```
The last release of `codeowners-validator` seems to have been in 2022. For longevity of this solution, it may be worthwhile to consider a fork, simply to ensure it gets regular updates, and we could include patches such as [the one you opened](https://github.com/mszostok/codeowners-validator/pull/222) without waiting for an unresponsive upstream

SuperSandro2000 replied:
Also it should set pname+version

cafkafk commented on .github/CODEOWNERS:
```diff
@@ -15,6 +15,8 @@
+/.github/workflows/codeowners.yml @infinisil
+/.github/OWNERS @infinisil
 /ci @infinisil @philiptaron @NixOS/Security
 
```
> This PR still runs the native CODEOWNERS together with the alternative mechanism, so that we can run it for some time to confirm that it works correctly. Once confirmed, we'll be able to easily turn off the native CODEOWNERS and just rely on the alternate mechanism.

I think it may still be worthwhile to keep the native codeowners file around for specific parts of the codebase, that way we can limit who can merge using branch rules on specific parts that could be considered critical.

Of course that kind usage should probably be limited, and wouldn't help with avoiding pings for those parts.

cafkafk replied:
Also, for what it's worth, validate codeowners could probably be contorted slightly to support failing on certain files we specify, and then have a CI step that could be required with a GitHub branch rule to block merge, but that would then need to be overridden either by someone with correct permissions or some more complicated mechanism.

infinisil replied:
Yeah there's a bunch more things we could do here. In particular I'm very interested in implementing local declaration of code ownership, e.g. using

```nix
/* Owner: @infinisil */
null
```

This could then be extended to support specifying required approvals or more:

```nix
/*
  Owner: @infinisil
  NeedsToApprove: @cafkafk 
*/
null
```

Or we could even do something like: Request a review by the maintainer, but if they don't respond within a week, request a review by the maintainer of the parent directory, etc.

Of course this all needs to be implemented unless we can find a third-party project that did it already, but the possibilities are almost endless :)

winterqt replied:
```
/*
  Owner: @infinisil
  NeedsToApprove: @cafkafk 
*/
```

To workshop a bit more: why would an NTA not be an owner and vice-versa?

infinisil replied:
I'd say one would implicitly be added to the other. I guess expressing it like
```
/*
  Owners: @infinisil @cafkafk (approval-needed)
*/
```

would make this less confusing

cafkafk reviewed (approved):
This looks good to me, thanks for working on this. I think that aside from just avoiding the ping spam it's gonna be very useful for making maintainers aware of changes to their packages that can often slip past them. I can't wait to try this on some of the many k8s related paths in nixpkgs :p

---

Maybe this should be put into draft status until https://github.com/NixOS/org/issues/31 has been solved? It's probably the easiest way to avoid an accidental merge.

winterqt commented:
I'm going to preemptively mark this as a draft to avoid, er, *"overly-eager"* pushes of the big green button.

Great work on this! (This isn't an approval, I've only skimmed the code (for now).)

infinisil commented:
GitHub App supposedly exists now, so this would be ready to merge

infinisil commented:
Also successfully tested it on the latest mass ping (https://github.com/NixOS/nixpkgs/pull/346433), seems to work!

```
$ nix-build ci -A requestReviews
$ result/bin/request-reviews.sh NixOS/nixpkgs 346433 .github/CODEOWNERS
[...]
The PR's base branch is set to master, but 5462 commits from the release-24.05 branch are included.
Make sure you know the right base branch for your changes: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#branch-conventions, then:
- If the changes should go to the release-24.05 branch instead, change the base branch accordingly:
  https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request
- If the changes should go to the master branch, rebase your PR onto merge base with the release-24.05 branch:
  git rebase --onto 0b49ef697b3e5cff1e6f18033de2b422e2fe1f43 6ec4e82869520bdc9ec93e9582e800182f6deb9c && git push --force-with-lease
```

winterqt commented on ci/request-reviews/verify-base-branch.sh:
```diff
@@ -0,0 +1,96 @@
+    # `headRef testBranch` with the merge base of the two.
+    extraCommits=$(git -C "$localRepo" rev-list --count "$testMergeBase" ^"$baseBranch")
+
+    if (( extraCommits != 0 )); then
+        log -en "\e[31m"
+        log "The PR's base branch is set to $baseBranch, but $extraCommits commits from the $testBranch branch are included."
+        log "Make sure you know the right base branch for your changes: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#branch-conventions, then:"
+        log "- If the changes should go to the $testBranch branch instead, change the base branch accordingly:"
+        log "  https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request"
+        log "- If the changes should go to the $baseBranch branch, rebase your PR onto merge base with the $testBranch branch:"
+        log "  git rebase --onto $prMergeBase $testMergeBase && git push --force-with-lease"
+        log -en "\e[0m"
+        exit 1
+    fi
```
Could we perhaps comment this on the PR instead, for better visibility? The user will probably get a "workflow failed" email, but meh.

(I assume you decided on this for a reason, but just in case you didn't think of it, figured I'd bring it up.)

infinisil replied:
I did think of it briefly. It does introduce more side effects: Comments could end up being posted multiple times for subsequent pushes, as each triggers a CI run. While requesting reviews is also a side-effect, it's an idempotent one, so it doesn't matter if it's done many times.

However thinking about this more deeply, I think the benefit of guiding newcomers in how to handle such scary situation smoothly outweighs the downside of an occasional double comment (if it even ever occurs), so let's do this :)

winterqt replied:
> Comments could end up being posted multiple times for subsequent pushes

We could do a GH API query and quickly grep for our (the app's) user ID, but now I'm trying to think if there'd be a situation where the message could be different per screw up down the line (e.g. user thinks they set to the right base branch, but then they screw up the rebase). Right now, this leads to the same comment, but what if we do some magic to get more intelligent remediation commands at some point? All's this to say that the double commenting may be a good thing, if my explanation makes sense.

infinisil replied:
There's actually some helpful reusable actions to help with this, and I've used this before [here](https://github.com/NixOS/nixfmt/blob/d0c0cda4fe6b7ae2b9aa37f3ee15e6412573eb01/.github/workflows/main.yml#L39-L94) ([in action](https://github.com/NixOS/nixfmt/pull/254#issuecomment-2356858761)).

But yeah agreed, I think it's fine to just not worry about the double comment for now. We can revisit if it becomes a problem.

winterqt replied:
Yup, sounds good.

infinisil replied:
(leaving this unresolved until I implement this)

infinisil replied:
Done and tested! https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/21

winterqt commented on ci/request-reviews/request-reviews.sh:
```diff
@@ -0,0 +1,76 @@
+log "Getting code owners to request reviews from"
+"$SCRIPT_DIR"/get-reviewers.sh "$tmp/nixpkgs.git" "$baseBranch" "$headRef" "$ownersFile" "$prAuthor" > "$tmp/reviewers.json"
+
+log "Requesting reviews from: $(<"$tmp/reviewers.json")"
```
Do we know if there a limit on how many reviewers can be added to a single PR, similar to description/comment ping limits?

infinisil replied:
The [API call](https://docs.github.com/en/rest/pulls/review-requests?apiVersion=2022-11-28#request-reviewers-for-a-pull-request) doesn't mention anything, and neither does https://github.com/dead-claudia/github-limits. The web interface says "Request up to 15 reviewers", but obviously we have many more for the mass pings. I guess we'll just see it when it happens, e.g. once https://github.com/NixOS/nixpkgs/pull/322537 is undrafted

winterqt commented on ci/codeowners-validator/default.nix:
```diff
@@ -0,0 +1,31 @@
+      url = "https://github.com/mszostok/codeowners-validator/compare/f3651e3810802a37bd965e6a9a7210728179d076...840eeb88b4da92bda3e13c838f67f6540b9e8529.patch";
+      hash = "sha256-t3Dtt8SP9nbO3gBrM0nRE7+G6N/ZIaczDyVHYAG/6mU=";
+    })
+    # Undoes part of the above PR: We don't want to require write access
+    # to the repository, that's only needed for GitHub's native CODEOWNERS.
+    # Furthermore, it removes an unneccessary check from the code
+    # that breaks tokens generated for GitHub Apps.
+    ./permissions.patch
+    # Allows setting a custom CODEOWNERS path using the OWNERS_FILE env var
+    ./owners-file-name.patch
```
Is there a reason you didn't merge at least the permissions patch (e.g. behind a flag) in your original PR?

Likewise for a PR upstream to the file path thing, though it is less of a concern given there's no overlap between your open PR and that specific functionality.

infinisil replied:
Because the native CODEOWNERS requires people to have write access, the permission patch removing that check only makes sense when you don't use the native CODEOWNERS. The original purpose of the PR (https://github.com/mszostok/codeowners-validator/pull/222) was really just do do that write check, but here I'm mainly using it for just the side benefits of improved performance:

> Speeds up the users check by not querying for all organisation members (gets just a bit slow [if you have 3k members](https://github.com/orgs/NixOS/people)!), instead just query the users with access to the repository

infinisil replied:
> Likewise for a PR upstream to the file path thing, though it is less of a concern given there's no overlap between your open PR and that specific functionality.

That would be good, I'm just a bit lazy :sweat_smile:

winterqt replied:
Makes sense, thanks!

winterqt reviewed (approved):
Overall, this is great! Just a few questions.

winterqt commented on ci/request-reviews/request-reviews.sh:
```diff
@@ -0,0 +1,76 @@
+log "Fetching PR info"
+prInfo=$(gh api \
+  -H "Accept: application/vnd.github+json" \
+  -H "X-GitHub-Api-Version: 2022-11-28" \
```
Worth turning this into a helper function (given the same API version header is used below) and/or document why we need to specify API version?

infinisil replied:
This is how the API calls are shown in the docs (e.g. [here](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request)). I guess this might be intentional, because if a new version is introduced, keeping the version for each call allows you to use the new version without having to update all of them at once.

I don't think it's worth abstracting this either way though, since there's [less than three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)).

infinisil commented:
Also added a dry mode and enabled it for now, so that no actual reviews get requested and no comments are posted while we're still trying this out

philiptaron commented on .github/workflows/codeowners.yml:
```diff
@@ -0,0 +1,88 @@
+    name: Request
+    runs-on: ubuntu-latest
+    # Don't trigger on draft PRs
+    if: ${{ ! github.event.pull_request.draft }}
```
Future: I'd love it if this left a review comment telling me who it _would_ add, because the "🌙" icon thing from the default GitHub UI will be disabled.

infinisil replied:
Ohh, actually that's _super_ simple with the `DRY_MODE` being a thing now! I'll give it a try :D

philiptaron commented on .github/workflows/codeowners.yml:
```diff
@@ -0,0 +1,88 @@
+    steps:
+    - uses: cachix/install-nix-action@9f70348d77d0422624097c4b7a75563948901306 # v29
+
+    # Important: Because we use pull_request_target, this checks out the base branch of the PR, not the PR head.
+    # This is intentional, because we need to request the review of owners as declared in the base branch.
```
It's really confusing that this is the default behavior!

infinisil replied:
Yeah there's a bunch of gotchas like this with GH Actions..

philiptaron commented on ci/request-reviews/default.nix:
```diff
@@ -0,0 +1,43 @@
+      ./get-reviewers.sh
+      ./request-reviews.sh
+      ./verify-base-branch.sh
+      ./devBranches.txt
```
Nit: If the other files in this directory use [`kebab-case`](https://en.wikipedia.org/wiki/Letter_case#Kebab_case), this one shouldn't use [`camelCase`](https://en.wikipedia.org/wiki/Camel_case).

philiptaron commented on ci/request-reviews/request-reviews.sh:
```diff
@@ -0,0 +1,95 @@
+# Fetch the PR
+git -C "$tmp/nixpkgs.git" remote add fork https://github.com/"$prRepo".git
+# Make sure we only fetch the commit history, nothing else
+git -C "$tmp/nixpkgs.git" config remote.fork.promisor true
```
Deep lore: https://git-scm.com/docs/git-config#Documentation/git-config.txt-remoteltnamegtpromisor

https://git-scm.com/docs/partial-clone/2.43.0 is what talks about promisor objects

infinisil replied:
Originally I copied these settings from how a remote is set up when you clone with `--filter=tree:0`, because I don't think it's possible to easily set up a second remote with them!

philiptaron commented on ci/request-reviews/request-reviews.sh:
```diff
@@ -0,0 +1,95 @@
+git -C "$tmp/nixpkgs.git" remote add fork https://github.com/"$prRepo".git
+# Make sure we only fetch the commit history, nothing else
+git -C "$tmp/nixpkgs.git" config remote.fork.promisor true
+git -C "$tmp/nixpkgs.git" config remote.fork.partialclonefilter tree:0
```
https://git-scm.com/docs/git-config#Documentation/git-config.txt-remoteltnamegtpartialclonefilter

I couldn't find straight-forward documentation of `tree:0` in the Git documentation, but I did find this tutorial on GitHub that refers to `tree:0` as a "tree-less clone".

https://github.blog/open-source/git/get-up-to-speed-with-partial-clone-and-shallow-clone/#user-content-quick-summary

infinisil replied:
Yeah that's also where I learned about it, I can add a comment up top :)

philiptaron commented on ci/request-reviews/verify-base-branch.sh:
```diff
@@ -0,0 +1,97 @@
+git -C "$localRepo" branch --list --format "%(refname:short)" "${devBranchPatterns[@]}" > "$tmp/dev-branches"
+readarray -t devBranches < "$tmp/dev-branches"
+
+if [[ "$baseRepo" == "$prRepo" ]] && elementIn "$prBranch" "${devBranches[@]}"; then
+    log "This PR merges $prBranch into $baseBranch, no commit check necessary"
+    exit 0
+fi
```
Spicy!

philiptaron replied:
@emilazy -- I think this implements [your most recent request](https://github.com/NixOS/nixpkgs/pull/336261#issuecomment-2398282281), if I'm simulating its effect correctly.

emilazy replied:
Aha, perhaps it does. That would be nice, if so. Even opening and immediately self‐merging a merge conflict resolution PR would feel less dicey than pushing directly.

infinisil replied:
Nah it doesn't because this is only the sub-script that checks whether the base branch is correct. It still requests reviews right now, but it could easily be changed by just flipping that 0 to a 1!

infinisil replied:
Turns out it's a bit more complicated than that, because there's three different scenarios:
- Dev branch merge: Succeed but don't ping
- Wrong base branch: Fail and don't ping
- Normal: Succeed and ping

And the script would need more changes to deal with this, so I'd rather keep this out of this PR to minimise further changes.

emilazy replied:
Yeah, that sounds reasonable. I think the case to handle specially is where the PR `HEAD` is a merge commit between commits that are all already on upstream branches.

philiptaron commented on ci/request-reviews/verify-base-branch.sh:
```diff
@@ -0,0 +1,97 @@
+
+for testBranch in "${devBranches[@]}"; do
+
+    if [[ -z "$(git -C "$localRepo" rev-list -1 --since="1 month ago" "$testBranch")" ]]; then
+        log "Not checking $testBranch, was inactive for the last month"
+        continue
+    fi
```
Nice performance optimization!

philiptaron commented on ci/request-reviews/verify-base-branch.sh:
```diff
@@ -0,0 +1,97 @@
+
+    if (( extraCommits != 0 )); then
+        log -e "\e[33m"
+        echo "The PR's base branch is set to $baseBranch, but $extraCommits commits from the $testBranch branch are included. Make sure you know the [right base branch for your changes](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#branch-conventions), then:"
+        echo "- If the changes should go to the $testBranch branch, [change the base branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request) to $testBranch"
+        echo "- If the changes should go to the $baseBranch branch, rebase your PR onto the merge base with the $testBranch branch:"
+        echo "  \`\`\`"
+        echo "  git rebase --onto $prMergeBase $testMergeBase"
+        echo "  git push --force-with-lease"
+        echo "  \`\`\`"
+        log -e "\e[m"
```
I do wish this were a comment on the PR, in addition to the failing action.

infinisil replied:
There is already! https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/21#issuecomment-2394800901

philiptaron reviewed (approved):
Approved with comments. I think this is a massive step forward in friendliness for new contributors, in addition to its other benefits for existing committers plagued by mass pings.

infinisil commented on ci/request-reviews/devBranches.txt:
```diff
@@ -0,0 +1,5 @@
+master
```
It's directly passed as arguments to `git branch --list`, so unfortunately not without extra processing, but that could be implemented, I'll give it a try

emilazy commented:
Something for future work: right now we push manual `master` → `staging-next` → `staging` merges directly to those branches, because opening a PR for them includes all the commits being merged in, causing mass pings in miniature. Similarly, the `staging-next` → `master` pull requests ping a lot of people for questionable value. It’d be nice if this machinery could detect such merge PRs and avoid pinging all the code owners.

infinisil commented:
Now addressed your comments @philiptaron: ([force push diff](https://github.com/NixOS/nixpkgs/compare/0e7b2ce40ecea70148f18dccea5538223b566074..93dcd42f2b208390491099ec5db481e31a47097e))

With the number of approvals and this not being set live just yet, I feel confident to merge this myself now, but I'd appreciate it if somebody else pushes the button :)

philiptaron reviewed (approved):
Thanks, changes look good, let's roll 🚀

philiptaron commented:
I pushed the button.

infinisil commented:
I briefly disabled it manually because apparently the current codeowners file references some non-existent files: https://github.com/NixOS/nixpkgs/actions/runs/11243668280/job/31260095472#step:7:34

infinisil commented:
Kind of dangerous though, that we can disable the workflow to prevent pings (well not yet), we'll have to trust the committers with that..

infinisil commented:
https://github.com/NixOS/nixpkgs/pull/347348

infinisil commented:
Alright merged and enabled again. Here's the place to watch how it's doing: https://github.com/NixOS/nixpkgs/actions/workflows/codeowners.yml

infinisil commented:
Another follow-up improvement: https://github.com/NixOS/nixpkgs/pull/347592

infinisil commented:
We also did have a mass ping earlier with https://github.com/NixOS/nixpkgs/pull/346556, but CI didn't get to the decisive part because the PR was closed before then. Instead it failed because the `/merge` ref didn't exist anymore: https://github.com/NixOS/nixpkgs/actions/runs/11261666327/job/31315601456#step:6:59. I'm looking into if it's possible to not fail like that

infinisil commented:
https://github.com/NixOS/nixpkgs/pull/347610 :rocket: 

We could also wait longer, but honestly I don't think we need to. We can always revert if it causes problems, but I'm fairly convinced there won't be any, and I'll be available in case there are

SuperSandro2000 commented:
We don't have any tests for this other than running it in dry-run mode?
That is kinda concerning for security related tool that talks to outside APIs and does many things in shellscript.

sternenseemann commented on ci/request-reviews/get-reviewers.sh:
```diff
@@ -0,0 +1,87 @@
+tmp=$(mktemp -d)
+trap 'rm -rf "$tmp"' exit
+
+git -C "$gitRepo" diff --name-only --merge-base "$baseRef" "$headRef" > "$tmp/touched-files"
```
This is not sound and apparently behaves differently to plain `git merge-base`, e.g. `git diff --name-only --merge-base origin/staging origin/haskell-updates` fails while you can compute `git merge-base origin/staging origin/haskell-updates`.

As a result, codeowners is currently send me an email on push as a service on haskell-updates.

infinisil replied:
Hmm I cannot reproduce a `git diff` failure with the current branch heads, can you give me specific commits that caused that or a link to a failed run?

sternenseemann replied:
Yes, sorry. E.g.: `git diff --name-only --merge-base df755a5e9c23fc1d9b5d801569e516889b28aec4 ea13a8bd5db362f54cce6a23178307217204e718` which should roughyl equate to [this run](https://github.com/NixOS/nixpkgs/actions/runs/12835006976/job/35793501215).
